File: systems-for-design-and-implementation/labs/exam/TicTacToeNetworking/src/tictactoe/network/rpcprotocol/Response.java
package tictactoe.network.rpcprotocol;

import java.io.Serializable;

public class Response implements Serializable
{
    private ResponseType type;
    private Object       data;

    private Response()
    {
    }

    public ResponseType getType()
    {
        return type;
    }

    private void setType(ResponseType type)
    {
        this.type = type;
    }

    public Object getData()
    {
        return data;
    }

    private void setData(Object data)
    {
        this.data = data;
    }

    @Override
    public String toString()
    {
        return "Response{" +
            "setType='" + type + '\'' +
            ", getData='" + data + '\'' +
            '}';
    }

    public static class Builder
    {
        private Response response = new Response();

        public Builder type(ResponseType type)
        {
            response.setType(type);
            return this;
        }

        public Builder data(Object data)
        {
            response.setData(data);
            return this;
        }

        public Response build()
        {
            return response;
        }
    }
}

File: systems-for-design-and-implementation/labs/lab3/LibraryNetworking/src/library/network/rpcprotocol/Response.scala
package library.network.rpcprotocol

import java.io.Serializable

import library.network.rpcprotocol.ResponseType.ResponseType

object Response
{

    class Builder
    {
        private val response: Response = new Response

        def `type`(`type`: String): Response.Builder =
        {
            response.setType(`type`)
            this
        }

        def data(data: Any): Response.Builder =
        {
            response.setData(data)
            this
        }

        def build: Response =
        {
            response
        }
    }

}

class Response private extends Serializable
{
    private var `type`: String = null
    private var data: Any = null

    def getType: String =
    {
        `type`
    }

    private def setType(`type`: String)
    {
        this.`type` = `type`
    }

    def getData: Any =
    {
        data
    }

    private def setData(data: Any)
    {
        this.data = data
    }

    override def toString: String =
    {
        "Response{" + "setType='" + `type` + '\'' + ", getData='" + data + '\'' + '}'
    }
}
